### PR TITLE
Fix SIGABRT caused by uninitialized mutex

### PR DESCRIPTION
**Pull request by ethercflow**
Fixed bug where wrong use of uninitialized Mutex caused SIGABRT.

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

### CODE REVIEW

**Comment by ethercflow** [comment]
@Connor1996 @tabokie PTAL, thanks!

File: db/write_thread.h [hunk 1]
```diff
@@ -185,6 +185,7 @@ class WriteThread {
     CommitRequest* request;
     SequenceNumber sequence;  // the sequence number to use for the first key
     Status status;
+    std::mutex status_mu;
```
Comment (tabokie) [inline db/write_thread.h]: You should figure out another solution. Maybe spinlock? Each write creates a writer, this puts too much load on futex infrastructure (last I checked each futex creation insert to a global kernel map).

File: db/write_thread.h [hunk 2]
```diff
@@ -185,6 +185,22 @@ class WriteThread {
     CommitRequest* request;
     SequenceNumber sequence;  // the sequence number to use for the first key
     Status status;
```
Comment (tabokie) [inline db/write_thread.h]: Comment that this field is protected by status_lock instead of db_mutex.

File: db/write_thread.h [hunk 3]
```diff
@@ -185,6 +185,22 @@ class WriteThread {
...
+        }
+        while (l.load(std::memory_order_relaxed)) {
+          port::AsmVolatilePause();
+        }
+      }
+
+      void unlock() noexcept {
+        l.store(false, std::memory_order_release);
+      }
+    } status_lock;
```
Comment (tabokie) [inline db/write_thread.h]: There's already a spinlock implementation in `util/mutexlock.h/SpinMutex`

**Review by tabokie — COMMENTED** [review]
Needs to add a new method for reading status. There're a lot of places that read status using db_mutex.

**Comment by ethercflow** [comment]
> Needs to add a new method for reading status. There're a lot of places that read status using db_mutex.

I don't think it's needed because `WriterThread` returns status only after writer has finished committing (by which time helpers have also finished modifying the `status` for this writer)

File: db/write_thread.h [hunk 4]
```diff
@@ -184,7 +185,8 @@ class WriteThread {
     WriteGroup* write_group;
     CommitRequest* request;
     SequenceNumber sequence;  // the sequence number to use for the first key
-    Status status;
+    Status status; // this field is protected by status_lock instead of db_mutex
+    SpinMutex status_lock;
```
Comment (tabokie) [inline db/write_thread.h]:
```suggestion
    Status status;  // write protected by status_lock in multi batch write.
    SpinMutex status_lock;
```

**Review by tabokie — APPROVED** [review]

**Review by Connor1996 — COMMENTED** [review]
why the mutex is uninitiliazed

**Comment by ethercflow** [comment]
> why the mutex is uninitiliazed

As this `mutex` is lazily created in [BlockingAwaitState](https://github.com/tikv/rocksdb/blob/6.4.tikv/db/write_thread.cc#L35), only when the new writer enters `JoinBatchGroup` and [not linked as the leader](https://github.com/tikv/rocksdb/blob/6.4.tikv/db/write_thread.cc#L395).

**Comment by Connor1996** [comment]
> > why the mutex is uninitiliazed
> 
> As this `mutex` is lazily created in [BlockingAwaitState](https://github.com/tikv/rocksdb/blob/6.4.tikv/db/write_thread.cc#L35), only when the new writer enters `JoinBatchGroup` and [not linked as the leader](https://github.com/tikv/rocksdb/blob/6.4.tikv/db/write_thread.cc#L395).

Not deep dive into the code, can we call `CreateMutex` somewhere instead of introducing a new mutex

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**